panic in dns resolver: invalid memory address or nil pointer dereference

I just upgraded to Docker 1.11.0 and noticed that the docker daemon crashes when I run a suite of tests that do the following:

1. create a network
2. start several containers with above network
3. run some tests
4. tear down containers and network

Each tests is run in parallel.

After a few iterations, I can reliably get the Docker daemon to crash with the following panic:


```
time="2016-04-15T10:50:03.670207907-06:00" level=debug msg="client dns id 12960, changed id 23756"
time="2016-04-15T10:50:03.670386643-06:00" level=debug msg="Can't retrieve client context for dns id 23227"
time="2016-04-15T10:50:03.670445392-06:00" level=debug msg="Query node-1-5_ebaf3361161b41c796c6106de40d0e7e.example.com.[28] from 172.21.0.4:41501, forwarding to udp:128.222.208.98"
panic: runtime error: invalid memory address or nil pointer dereference         
[signal 0xb code=0x1 addr=0x30 pc=0x87d73d]                                     
                                                                                
goroutine 105572 [running]:                                                     
github.com/docker/libnetwork.(*resolver).forwardQueryStart(0xc82230d700, 0x0, 0x0, 0xc8225cc240, 0x0)
    /usr/src/docker/vendor/src/github.com/docker/libnetwork/resolver.go:442 +0x5d
github.com/docker/libnetwork.(*resolver).ServeDNS(0xc82230d700, 0x0, 0x0, 0xc8225cc240)
    /usr/src/docker/vendor/src/github.com/docker/libnetwork/resolver.go:391 +0xde4
github.com/miekg/dns.(*Server).serve(0xc820340dd0, 0x7fb230734f18, 0xc821b1cab0, 0x7fb230734fc0, 0xc82230d700, 0xc82099c400, 0x51, 0x200, 0xc820339d48, 0xc820d30e00, ...)
    /usr/src/docker/vendor/src/github.com/miekg/dns/server.go:535 +0x7c1        
created by github.com/miekg/dns.(*Server).serveUDP                              
    /usr/src/docker/vendor/src/github.com/miekg/dns/server.go:489 +0x3d5        
```

--- a/resolver.go
+++ b/resolver.go
@@ -305,6 +305,7 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 		extConn net.Conn
 		resp    *dns.Msg
 		err     error
+		writer  dns.ResponseWriter
 	)
 
 	if query == nil || len(query.Question) == 0 {
@@ -342,7 +343,9 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 		if resp.Len() > maxSize {
 			truncateResp(resp, maxSize, proto == "tcp")
 		}
+		writer = w
 	} else {
+		queryID := query.Id
 		for i := 0; i < maxExtDNS; i++ {
 			extDNS := &r.extDNSList[i]
 			if extDNS.ipStr == "" {
@@ -388,7 +391,7 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 
 			// forwardQueryStart stores required context to mux multiple client queries over
 			// one connection; and limits the number of outstanding concurrent queries.
-			if r.forwardQueryStart(w, query) == false {
+			if r.forwardQueryStart(w, query, queryID) == false {
 				old := r.tStamp
 				r.tStamp = time.Now()
 				if r.tStamp.Sub(old) > logInterval {
@@ -418,32 +421,33 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 
 			// Retrieves the context for the forwarded query and returns the client connection
 			// to send the reply to
-			w = r.forwardQueryEnd(w, resp)
-			if w == nil {
+			writer = r.forwardQueryEnd(w, resp)
+			if writer == nil {
 				continue
 			}
 
 			resp.Compress = true
 			break
 		}
-
-		if resp == nil || w == nil {
+		if resp == nil || writer == nil {
 			return
 		}
 	}
 
-	err = w.WriteMsg(resp)
-	if err != nil {
+	if writer == nil {
+		return
+	}
+	if err = writer.WriteMsg(resp); err != nil {
 		log.Errorf("error writing resolver resp, %s", err)
 	}
 }
 
-func (r *resolver) forwardQueryStart(w dns.ResponseWriter, msg *dns.Msg) bool {
+func (r *resolver) forwardQueryStart(w dns.ResponseWriter, msg *dns.Msg, queryID uint16) bool {
 	proto := w.LocalAddr().Network()
 	dnsID := uint16(rand.Intn(maxDNSID))
 
 	cc := clientConn{
-		dnsID:      msg.Id,
+		dnsID:      queryID,
 		respWriter: w,
 	}
 

